rust/gui-client/src-tauri/src/client/gui.rs

```diff
@@ -452,10 +452,27 @@ pub(crate) enum ControllerRequest {
 enum Status {
     /// Firezone is disconnected.
     Disconnected,
-    /// Firezone is signing in and raising the tunnel.
-    Connecting { start_instant: Instant },
     /// Firezone is ready to use.
     TunnelReady { resources: Vec<ResourceDescription> },
+    /// At least one connection request has failed, due to failing to reach the Portal, and we are waiting for a network change before we try again
+    WaitingForNetwork {
+        /// The token to log in to the Portal, for retrying the connection request.
+        token: SecretString,
+    },
+    /// Firezone is signing in to the Portal.
+    WaitingForPortal {
+        /// True if this is user-initiated and we haven't done any automatic retries yet.
+        first_time: bool,
+        /// The instant when we sent our most recent connect request.
+        start_instant: Instant,
+        /// The token to log in to the Portal, in case we need to retry the connection request.
+        token: SecretString,
+    },
+    /// Firezone has connected to the Portal and is raising the tunnel.
+    WaitingForTunnel {
+        /// The instant when we sent our most recent connect request.
+        start_instant: Instant,
+    },
```

Comment:
Can we perhaps merge some of these together and make the logic on what the GUI does a bit more general?

1. On start-up, try to sign in. Show to user as "Connecting".
2. If sign-in fails due to any reason that is not an authentication error, still show "Connecting" but internally move to a `ConnectFailed` state that holds the token.
3. On network changes, try to send another connect when in that state.
4. Optionally: Provide a way for the user to force another connect by maybe adding a "Retry now" option to the "Connecting" element in the GUI.

Reply:
So generalize `WaitingForNetwork` to `RetryingConnection`? 

I prefer naming states for either what their next step is ("Waiting for X") or at least what they're doing ("Retrying"), if it's `ConnectFailed` it doesn't say where it's going

Reply:
> I prefer naming states for either what their next step is ("Waiting for X") or at least what they're doing ("Retrying"), if it's `ConnectFailed` it doesn't say where it's going

More complex state machines will potentially:

- Transition to different states
- Offer multiple things that can be done

What they still have in common is that multiple transitions lead to the same state. In that case, naming it after what happened can be a good middle-ground (`SignedIn`, `ConnectFailed`).

I don't mind either way, `RetryingConnection` also works for me :)